### ISSUE TITLE
More examples needed.

### ISSUE THREAD

**Issue by hetii** [kadalu/kadalu#42]
Hi,
Could you provide more examples, what to do when we have more nodes, and how to configure kadalu to have at least one mirroring data in case of some crash?

General some descriptions regarding data recovery plan or redeploying kadalu would be useful.

Best regards.

**Comment by amarts** [comment]
Sure. Makes sense to add some documents on this regard for sure. Will do that soon.

Btw, will answer some of the above questions for you here itself:

#### 1. What do we do when we have more nodes?
Kadalu's design principles are with couple of assumptions: (ie, In the early development phase).

a. The PVC size requirements are going to be lesser than one available disk of any node.
b. No need to change settings once Gluster is running (in most of the cases).

With that assumption, lets look into what happens when you have more node. For easier to understand reasons, I am assuming `kadalu.replica1` mode, so there is 1 brick process per a disk (or available storage, like EBS, RAID device, SAN Array etc).

Now, the examples given are for setting up it on 1 node, 1 device. When you have more storage you need to export, then just export more storage with multiple storage config definitions.

For example:

```
# File: storage-config.yaml
---
apiVersion: kadalu-operator.storage/v1alpha1
kind: KadaluStorage
metadata:
  # This will be used as name of PV Hosting Volume
  name: storage-pool-1
spec:
  type: Replica1
  storage:
    - node: kube1      # node name as shown in `kubectl get nodes`
      device: /dev/vdc # Device to provide storage to all PV
---
apiVersion: kadalu-operator.storage/v1alpha1
kind: KadaluStorage
metadata:
  # This will be used as name of PV Hosting Volume
  name: storage-pool-2 # **Notice** the name change here
spec:
  type: Replica1
  storage:
    - node: kube2      # node name as shown in `kubectl get nodes`
      device: /dev/vdd # Device to provide storage to all PV
---
apiVersion: kadalu-operator.storage/v1alpha1
kind: KadaluStorage
metadata:
  # This will be used as name of PV Hosting Volume
  name: storage-pool-3 # **Notice** the name change here
spec:
  type: Replica1
  storage:
    - node: kube3       # node name as shown in `kubectl get nodes`
      device: /dev/vdc # Device to provide storage to all PV

```
Note that you just need to provide one more config with different name to host a brick process. Our CSI driver will automatically understand there are 3 volumes which it can connect to, and depending on the available space, will provide storage in one of the gluster volume.

The newer configs can be given at any time, and CSI driver will understand the existence of new storage, (through kadalu operator), and will be consumed when more PV requests come by.

#### 2. how to configure kadalu to have at least one mirroring data in case of some crash?

The answer we have is, providing kadalu-config using 3 nodes, and using gluster replicate module (replica 3). The sample looks something like below:

```
# File: storage-config.yaml
---
apiVersion: kadalu-operator.storage/v1alpha1
kind: KadaluStorage
metadata:
  # This will be used as name of PV Hosting Volume
  name: storage-replica-pool-1
spec:
  type: Replica3  # Notice that this field tells kadalu operator to use replicate module.
  storage:
    - node: kube1      # node name as shown in `kubectl get nodes`
      device: /dev/vdc # Device to provide storage to all PV
    - node: kube2      # node name as shown in `kubectl get nodes`
      device: /dev/vdd # Device to provide storage to all PV
    - node: kube3      # node name as shown in `kubectl get nodes`
      device: /dev/vdc # Device to provide storage to all PV
---
```

With this, there will be 3 bricks, and kadalu CSI driver will mount the corresponding volume and provide data.

Also note that, there can be both replica1 and replica3 type volume co-existing in the system. Note that while claiming the PV, you just need to provide `storageClassName: kadalu.replica1` or `storageClassName: kadalu.replica3` to use the relevant option.

#### 3. On data recovery

As we use glusterfs in backend, without any sharding/striping/disperse mode, the data remains as is, on your backend storage. Just that each PV would be a subdirectory on your storage. So, no need to panic.

#### 4. On upgrade

As long as glusterfs promises to keep the backend layout same, and continue to provide storage after upgrade, we don't see any issue with upgrade. Currently our operator is not checking for newer versions and upgrading itself, but just killing the brick pod when new version is available should fetch new pod and start it back all fine.

--------
Hope this helps to get some answers.

**Comment by hetii** [comment]
Hello. Thank you for your explanation.

I love kadalu simplicity but have more questions.

Currently in your example you use the block devices. Is it possible to use directory?
If yes, then what about fs type need to be used for it?

I wonder also if kadalu can act as a middleware between PV/PVC delivered by cloud provider that only support RWO mode and use it as a backend storage for glusterfs who will offer it in RWX mode.

In my particular case I try to have RWX mode in azure aks, that only offer PVC in RWO (they support RWX by using crappy SAMBA protocol that I want to avoid).

**Comment by amarts** [comment]
Happy to answer.

> Currently in your example you use the block devices. Is it possible to use directory?

Yes :+1: 

> If yes, then what about fs type need to be used for it?

Any FS which supports xattr. We have tested mostly with xfs. May be for other FS type, we may need more testing.

Option looks something like:

```
# File: storage-config.yaml
---
apiVersion: kadalu-operator.storage/v1alpha1
kind: KadaluStorage
metadata:
  # This will be used as name of PV Hosting Volume
  name: storage-pool-1
spec:
  type: Replica1
  storage:
    - node: kube1               # node name as shown in `kubectl get nodes`
      path: /my/export/path # path to provide storage to all PV
```

> I wonder also if kadalu can act as a middleware between PV/PVC delivered by cloud provider that only support RWO mode and use it as a backend storage for glusterfs who will offer it in RWX mode.

Again, technically possible. Few more lines of code needed to be added in kadalu-server.

Option can be something like:

```
# File: storage-config.yaml
---
apiVersion: kadalu-operator.storage/v1alpha1
kind: KadaluStorage
metadata:
  # This will be used as name of PV Hosting Volume
  name: storage-pool-1
spec:
  type: Replica1
  storage:
    - node: kube1
      pvc:  cloud-provider-pvc # PVC name to use in the container
```
Here, "node" name is where it would be starting kadalu-server. Can be expanded in future to be handled by operator based on some tag, so it can decide where to start, and if the node is not there, it can start same exporter on other node, based on tag.

> In my particular case I try to have RWX mode in azure aks, that only offer PVC in RWO (they support RWX by using crappy SAMBA protocol that I want to avoid).

For now, this is not supported out of the box, but happy to get some help :-)

**Comment by aravindavk** [comment]
Will RWO PV created from Azure aks supports providing mount option and Filesystem type?

If RWO PV is available, (volumeMode: Filesystem)

- CSI driver of respective RWO provider should support setting mount option(In our case `-oprjquota` is required)
- Option to specify Filesystem type is required. Kadalu is using `xfs` to support directory Quota.

If Raw block volume(https://kubernetes.io/docs/concepts/storage/persistent-volumes/#raw-block-volume-support) is available,  (volumeMode: Block)

Supporting this is relatively easy compared to the previous option. The server pod gets a raw device, it can mount as required.

**Comment by amarts** [comment]
Closed with #50

**Comment by hetii** [comment]
I just add that in azure it's possible to use xfs fsType:

    ---
    kind: StorageClass
    apiVersion: storage.k8s.io/v1
    metadata:
      name: premium-azure-disk-sc
    provisioner: kubernetes.io/azure-disk
    reclaimPolicy: Retain
    # reclaimPolicy: Delete
    allowVolumeExpansion: true
    volumeBindingMode: Immediate
    parameters:
      storageaccounttype: Premium_LRS
      kind: Managed
      fsType: xfs
      #kind: Shared
